api/analyticsd/service/service.go

```diff
@@ -0,0 +1,200 @@
+}
+
+var _ pb.AnalyticsServiceServer = (*Service)(nil)
+
+type Config struct {
+	Listener       net.Listener
+	SegmentAPIKey  string
+	SegmentPrefix  string
+	FilrewardsAddr string
+	Debug          bool
+}
+
+func New(conf Config) (*Service, error) {
+	if conf.Debug {
+		if err := util.SetLogLevels(map[string]logging.LogLevel{
+			"analytics": logging.LevelDebug,
+		}); err != nil {
+			return nil, fmt.Errorf("setting log levels: %v", err)
+		}
+	}
+
+	config := segment.Config{
+		Verbose: conf.Debug,
+	}
+	segment, err := segment.NewWithConfig(conf.SegmentAPIKey, config)
+	if err != nil {
+		return nil, fmt.Errorf("creating segment client: %v", err)
+	}
+
+	s := &Service{
+		segment: segment,
+		prefix:  conf.SegmentPrefix,
+		debug:   conf.Debug,
+	}
+
+	if conf.FilrewardsAddr != "" {
+		s.frc, err = filrewards.New(conf.FilrewardsAddr)
+		if err != nil {
+			return nil, fmt.Errorf("creating filrewards client: %v", err)
+		}
+	}
+
+	s.server = grpc.NewServer()
+	go func() {
+		pb.RegisterAnalyticsServiceServer(s.server, s)
+		if err := s.server.Serve(conf.Listener); err != nil && !errors.Is(err, grpc.ErrServerStopped) {
+			log.Errorf("serve error: %v", err)
+		}
+	}()
+
+	return s, nil
+}
+
+func (s *Service) Identify(ctx context.Context, req *pb.IdentifyRequest) (*pb.IdentifyResponse, error) {
+	if req.AccountType == 2 {
+		return &pb.IdentifyResponse{}, nil // Or should we return an error?
+	}
+	traits := segment.NewTraits()
+	traits.Set("account_type", req.AccountType)
+	traits.Set(s.prefix+"signup", "true")
+	if req.Email != "" {
+		traits.SetEmail(req.Email)
+	}
+	for key, value := range req.Properties.AsMap() {
+		traits.Set(key, value)
+	}
+	err := s.segment.Enqueue(segment.Identify{
+		UserId: req.Key,
+		Traits: traits,
+		Context: &segment.Context{
+			Extra: map[string]interface{}{
+				"active": req.Active,
+			},
+		},
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "enqueuing segment identify: %v", err)
+	}
+	return &pb.IdentifyResponse{}, nil
+}
+
+func (s *Service) Track(ctx context.Context, req *pb.TrackRequest) (*pb.TrackResponse, error) {
+	if req.AccountType == 2 {
+		return &pb.TrackResponse{}, nil // Or should we return an error?
+	}
+	props := segment.NewProperties()
+	for key, value := range req.Properties.AsMap() {
+		props.Set(key, value)
+	}
+
+	if err := s.segment.Enqueue(segment.Track{
+		UserId:     req.Key,
+		Event:      eventToString(req.Event),
+		Properties: props,
+		Context: &segment.Context{
+			Extra: map[string]interface{}{
+				"active": req.Active,
+			},
+		},
+	}); err != nil {
+		return nil, status.Errorf(codes.Internal, "enqueuing segment track: %v", err)
+	}
+
+	// Don't process the event that is generated by s.frc.ProcessAnalyticsEvent
+	// when a reward record is created and sent here for analytics tracking.
+	if s.frc != nil && req.Event != pb.Event_EVENT_FIL_REWARD_RECORDED {
+		if _, err := s.frc.ProcessAnalyticsEvent(ctx, req.Key, req.AccountType, req.Event); err != nil {
```

Comment:
Here, we send the analytics event to the filrewards service using the new `filrewards.Client`.

Reply:
only funny thing i see here is that if the segment queue barks for some reason the reward wont be unlocked.

Reply:
Yea, that is the issue with relying on segment. But technically we just use email for notifications, so the rewards can still show up in any other UI.